### PR TITLE
planner: fix wrong selection push down when having above agg

### LINKED ISSUE #26496: having clause does not work correctly
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
drop table if exists UK_NSPRE_19416;
CREATE TABLE `UK_NSPRE_19416` (  `COL1` binary(20) DEFAULT NULL,  `COL2` varchar(20) DEFAULT NULL,  `COL4` datetime DEFAULT NULL,  `COL3` bigint(20) DEFAULT NULL,  `COL5` float DEFAULT NULL,  UNIQUE KEY `UK_COL1` (`COL1`)) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
insert into `UK_NSPRE_19416`(col1) values (0xc5b428e2ebc1b78f0b183899a8df55c88a333f86), (0x004dad637b37cc4a9742484ab93f97ede2ab8bd5), (0x550c4a4390ba14fd6d382dd29063e10210c99381);
select t1.col1, count(t2.col1) from UK_NSPRE_19416 as t1 left join UK_NSPRE_19416 as t2 on t1.col1 = t2.col1 where t1.col1 in (0x550C4A4390BA14FD6D382DD29063E10210C99381, 0x004DAD637B37CC4A9742484AB93F97EDE2AB8BD5, 0xC5B428E2EBC1B78F0B183899A8DF55C88A333F86) group by t1.col1, t2.col1 having t1.col1 in (0x9B4B48FEBA9225BACF8F9ADEAEE810AEC26DC7A2, 0x25A6C4FAD832F8E0267AAA504CFAE767565C8B84, 0xE26E5B0080EC5A8156DACE67D13B239500E540E6);
```

### 2. What did you expect to see? (Required)
```sql
mysql> select t1.col1, count(t2.col1) from UK_NSPRE_19416 as t1 left join UK_NSPRE_19416 as t2 on t1.col1 = t2.col1 where t1.col1 in (0x550C4A4390BA14FD6D382DD29063E10210C99381, 0x004DAD637B37CC4A9742484AB93F97EDE2AB8BD5, 0xC5B428E2EBC1B78F0B183899A8DF55C88A333F86) group by t1.col1, t2.col1 having t1.col1 in (0x9B4B48FEBA9225BACF8F9ADEAEE810AEC26DC7A2, 0x25A6C4FAD832F8E0267AAA504CFAE767565C8B84, 0xE26E5B0080EC5A8156DACE67D13B239500E540E6);
Empty set (0.06 sec)
```
### 3. What did you see instead (Required)
4.0.14
```sql
mysql> select t1.col1, count(t2.col1) from UK_NSPRE_19416 as t1 left join UK_NSPRE_19416 as t2 on t1.col1 = t2.col1 where t1.col1 in (0x550C4A4390BA14FD6D382DD29063E10210C99381, 0x004DAD637B37CC4A9742484AB93F97EDE2AB8BD5, 0xC5B428E2EBC1B78F0B183899A8DF55C88A333F86) group by t1.col1, t2.col1 having t1.col1 in (0x9B4B48FEBA9225BACF8F9ADEAEE810AEC26DC7A2, 0x25A6C4FAD832F8E0267AAA504CFAE767565C8B84, 0xE26E5B0080EC5A8156DACE67D13B239500E540E6);
+--------------------------------------------+----------------+
| col1                                       | count(t2.col1) |
+--------------------------------------------+----------------+
| 0x550C4A4390BA14FD6D382DD29063E10210C99381 |              1 |
| 0xC5B428E2EBC1B78F0B183899A8DF55C88A333F86 |              1 |
+--------------------------------------------+----------------+
```

5.0.3 & 5.1.0 & master(4116dd248a1533c83eebd4eaacc)
```sql
MySQL > select t1.col1, count(t2.col1) from UK_NSPRE_19416 as t1 left join UK_NSPRE_19416 as t2 on t1.col1 = t2.col1 where t1.col1 in (0x550C4A4390BA14FD6D382DD29063E10210C99381, 0x004DAD637B37CC4A9742484AB93F97EDE2AB8BD5, 0xC5B428E2EBC1B78F0B183899A8DF55C88A333F86) group by t1.col1, t2.col1 having t1.col1 in (0x9B4B48FEBA9225BACF8F9ADEAEE810AEC26DC7A2, 0x25A6C4FAD832F8E0267AAA504CFAE767565C8B84, 0xE26E5B0080EC5A8156DACE67D13B239500E540E6);
+--------------------------------------------+----------------+
| col1                                       | count(t2.col1) |
+--------------------------------------------+----------------+
| 0xc5b428e2ebc1b78f0b183899a8df55c88a333f86 | 1              |
| 0x004dad637b37cc4a9742484ab93f97ede2ab8bd5 | 1              |
| 0x550c4a4390ba14fd6d382dd29063e10210c99381 | 1              |
+--------------------------------------------+----------------+
```

### PR DESCRIPTION
**Pull request by lzmhhh123**
Signed-off-by: lzmhhh123 <lzmhhh123@gmail.com>

### What problem does this PR solve?

Issue Number: close #26496

### What is changed and how it works?

How it Works:

### Check List 

Tests 

- [X] Unit test
- [X] Integration test

Side effects

- [X] Performance regression: Consumes more CPU

### Release note 

```release-note
fix wrong selection push down when having above agg
```

### CODE REVIEW

File: planner/core/testdata/ordered_result_mode_suite_out.json [hunk 1]
```diff
@@ -417,12 +417,12 @@
       },
       {
         "Plan": [
-          "Sort_9 6400.00 root  Column#5, Column#6, Column#7",
-          "└─Selection_11 6400.00 root  lt(Column#6, 20)",
-          "  └─HashAgg_16 8000.00 root  group by:test.t1.d, funcs:min(Column#11)->Column#5, funcs:max(Column#12)->Column#6, funcs:sum(Column#13)->Column#7",
-          "    └─TableReader_17 8000.00 root  data:HashAgg_12",
-          "      └─HashAgg_12 8000.00 cop[tikv]  group by:test.t1.d, funcs:min(test.t1.a)->Column#11, funcs:max(test.t1.b)->Column#12, funcs:sum(test.t1.c)->Column#13",
-          "        └─TableFullScan_15 10000.00 cop[tikv] table:t1 keep order:false, stats:pseudo"
+          "Selection_8 6400.00 root  lt(Column#6, 20)",
```
Comment (wshwsh12) [inline planner/core/testdata/ordered_result_mode_suite_out.json:420]: Why the selection_8 can't push down to sort?

**Review by guo-shaoge — APPROVED** [review]

File: planner/core/testdata/ordered_result_mode_suite_out.json (hunk 1 above)
Comment (lzmhhh123) [reply planner/core/testdata/ordered_result_mode_suite_out.json:420]: It's a tradeoff. For the correct result. Actually, it can be pushed down to sort.

**Review by wshwsh12 — APPROVED** [review]
LGTM.  Maybe we need create another issue to track the plan regression.

**Pull request merged**